// noflo/lib/nodejs/ComponentLoader.js
(function() {
  var CoffeeScript, ComponentLoader, babel, fs, internalSocket, loader, log, nofloGraph, path, reader, utils, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  reader = require('read-installed');

  _ = require('underscore')._;

  path = require('path');

  fs = require('fs');

  loader = require('../ComponentLoader');

  internalSocket = require('../InternalSocket');

  utils = require('../Utils');

  nofloGraph = require('../Graph');

  CoffeeScript = require('coffee-script');

  if (typeof CoffeeScript.register !== 'undefined') {
    CoffeeScript.register();
  }

  babel = require('babel-core');

  log = require('npmlog');

  log.pause();

  ComponentLoader = (function(_super) {
    __extends(ComponentLoader, _super);

    function ComponentLoader() {
      return ComponentLoader.__super__.constructor.apply(this, arguments);
    }

    ComponentLoader.prototype.getModuleComponents = function(moduleDef, callback) {
      var checkOwn, components, depCount, done;
      components = {};
      this.checked.push(moduleDef.name);
      depCount = _.keys(moduleDef.dependencies).length;
      done = _.after(depCount + 1, (function(_this) {
        return function() {
          return callback(components);
        };
      })(this));
      _.each(moduleDef.dependencies, (function(_this) {
        return function(def) {
          if (def.name == null) {
            return done();
          }
          if (_this.checked.indexOf(def.name) !== -1) {
            return done();
          }
          return _this.getModuleComponents(def, function(depComponents) {
            var cPath, name;
            if (_.isEmpty(depComponents)) {
              return done();
            }
            for (name in depComponents) {
              cPath = depComponents[name];
              components[name] = cPath;
            }
            return done();
          });
        };
      })(this));
      if (!moduleDef.noflo) {
        return done();
      }
      checkOwn = (function(_this) {
        return function(def) {
          var cPath, gPath, loaderPath, name, prefix, _ref, _ref1;
          prefix = _this.getModulePrefix(def.name);
          if (def.noflo.icon) {
            _this.libraryIcons[prefix] = def.noflo.icon;
          }
          if (def.noflo.components) {
            _ref = def.noflo.components;
            for (name in _ref) {
              cPath = _ref[name];
              _this.registerComponent(prefix, name, path.resolve(def.realPath, cPath));
            }
          }
          if (moduleDef.noflo.graphs) {
            _ref1 = def.noflo.graphs;
            for (name in _ref1) {
              gPath = _ref1[name];
              _this.registerGraph(prefix, name, path.resolve(def.realPath, gPath));
            }
          }
          if (def.noflo.loader) {
            loaderPath = path.resolve(def.realPath, def.noflo.loader);
            if (!_this.componentLoaders) {
              _this.componentLoaders = [];
            }
            _this.componentLoaders.push(loaderPath);
            loader = require(loaderPath);
            return _this.registerLoader(loader, done);
          } else {
            return done();
          }
        };
      })(this);
      if (!this.revalidate) {
        return checkOwn(moduleDef);
      }
      return this.readPackageFile("" + moduleDef.realPath + "/package.json", function(err, data) {
        if (err) {
          return done();
        }
        return checkOwn(data);
      });
    };

    ComponentLoader.prototype.getCoreComponents = function(callback) {
      var corePath;
      corePath = path.resolve(__dirname, '../../src/components');
      if (path.extname(__filename) === '.coffee') {
        corePath = path.resolve(__dirname, '../../components');
      }
      return fs.readdir(corePath, (function(_this) {
        return function(err, components) {
          var component, componentExtension, componentName, coreComponents, _i, _len, _ref;
          coreComponents = {};
          if (err) {
            return callback(coreComponents);
          }
          for (_i = 0, _len = components.length; _i < _len; _i++) {
            component = components[_i];
            if (component.substr(0, 1) === '.') {
              continue;
            }
            _ref = component.split('.'), componentName = _ref[0], componentExtension = _ref[1];
            if (componentExtension !== 'coffee') {
              continue;
            }
            coreComponents[componentName] = "" + corePath + "/" + component;
          }
          return callback(coreComponents);
        };
      })(this));
    };

    ComponentLoader.prototype.cachePath = function() {
      return path.resolve(this.baseDir, './.noflo.json');
    };

    ComponentLoader.prototype.writeCache = function(callback) {
      var cPath, cacheData, filePath, loaders, name, _ref;
      cacheData = {
        components: {},
        loaders: []
      };
      loaders = this.componentLoaders || [];
      cacheData.loaders = loaders.map((function(_this) {
        return function(lPath) {
          return path.relative(_this.baseDir, lPath);
        };
      })(this));
      _ref = this.components;
      for (name in _ref) {
        cPath = _ref[name];
        if (typeof cPath !== 'string') {
          continue;
        }
        cacheData.components[name] = path.relative(this.baseDir, cPath);
      }
      filePath = this.cachePath();
      return fs.writeFile(filePath, JSON.stringify(cacheData, null, 2), {
        encoding: 'utf-8'
      }, callback);
    };

    ComponentLoader.prototype.readCache = function(callback) {
      var filePath;
      filePath = this.cachePath();
      return fs.readFile(filePath, {
        encoding: 'utf-8'
      }, (function(_this) {
        return function(err, cached) {
          var cPath, cacheData, done, e, name, _ref, _ref1;
          if (err) {
            return callback(err);
          }
          if (!cached) {
            return callback(new Error('No cached components found'));
          }
          try {
            cacheData = JSON.parse(cached);
          } catch (_error) {
            e = _error;
            callback(e);
          }
          if (!cacheData.components) {
            return callback(new Error('No components in cache'));
          }
          _this.components = {};
          _ref = cacheData.components;
          for (name in _ref) {
            cPath = _ref[name];
            _this.components[name] = path.resolve(_this.baseDir, cPath);
          }
          if (!((_ref1 = cacheData.loaders) != null ? _ref1.length : void 0)) {
            callback(null);
            return;
          }
          done = _.after(cacheData.loaders.length, function() {
            return callback(null);
          });
          return cacheData.loaders.forEach(function(loaderPath) {
            loader = require(path.resolve(_this.baseDir, loaderPath));
            return _this.registerLoader(loader, done);
          });
        };
      })(this));
    };

    ComponentLoader.prototype.listComponents = function(callback) {
      var done;
      if (this.processing) {
        this.once('ready', (function(_this) {
          return function() {
            return callback(_this.components);
          };
        })(this));
        return;
      }
      if (this.components) {
        return callback(this.components);
      }
      this.ready = false;
      this.processing = true;
      if (this.options.cache && !this.failedCache) {
        this.readCache((function(_this) {
          return function(err) {
            if (err) {
              _this.failedCache = true;
              _this.processing = false;
              _this.listComponents(callback);
              return;
            }
            _this.ready = true;
            _this.processing = false;
            _this.emit('ready', true);
            if (callback) {
              return callback(_this.components);
            }
          };
        })(this));
        return;
      }
      this.components = {};
      done = _.after(2, (function(_this) {
        return function() {
          _this.ready = true;
          _this.processing = false;
          _this.emit('ready', true);
          if (!_this.options.cache) {
            if (callback) {
              callback(_this.components);
            }
            return;
          }
          _this.writeCache(function() {
            if (callback) {
              return callback(_this.components);
            }
          });
        };
      })(this));
      this.getCoreComponents((function(_this) {
        return function(coreComponents) {
          var cPath, name;
          for (name in coreComponents) {
            cPath = coreComponents[name];
            _this.components[name] = cPath;
          }
          return done();
        };
      })(this));
      return reader(this.baseDir, (function(_this) {
        return function(err, data) {
          if (err) {
            return done();
          }
          return _this.getModuleComponents(data, function(components) {
            var cPath, name;
            for (name in components) {
              cPath = components[name];
              _this.components[name] = cPath;
            }
            return done();
          });
        };
      })(this));
    };

    ComponentLoader.prototype.getPackagePath = function(packageId, callback) {
      var find, found, seen;
      found = null;
      seen = [];
      find = function(packageData) {
        if (seen.indexOf(packageData.name) !== -1) {
          return;
        }
        seen.push(packageData.name);
        if (packageData.name === packageId) {
          found = path.resolve(packageData.realPath, './package.json');
          return;
        }
        return _.each(packageData.dependencies, find);
      };
      return reader(this.baseDir, function(err, data) {
        if (err) {
          return callback(err);
        }
        find(data);
        return callback(null, found);
      });
    };

    ComponentLoader.prototype.setSource = function(packageId, name, source, language, callback) {
      var Module, e, implementation, moduleImpl, modulePath;
      if (!this.ready) {
        this.listComponents((function(_this) {
          return function() {
            return _this.setSource(packageId, name, source, language, callback);
          };
        })(this));
        return;
      }
      Module = require('module');
      if (language === 'coffeescript') {
        try {
          source = CoffeeScript.compile(source, {
            bare: true
          });
        } catch (_error) {
          e = _error;
          return callback(e);
        }
      } else if (language === 'es6' || language === 'es2015') {
        try {
          source = babel.transform(source).code;
        } catch (_error) {
          e = _error;
          return callback(e);
        }
      }
      try {
        modulePath = path.resolve(this.baseDir, "./components/" + name + ".js");
        moduleImpl = new Module(modulePath, module);
        moduleImpl.paths = Module._nodeModulePaths(path.dirname(modulePath));
        moduleImpl.filename = modulePath;
        moduleImpl._compile(source, modulePath);
        implementation = moduleImpl.exports;
      } catch (_error) {
        e = _error;
        return callback(e);
      }
      if (!(implementation || implementation.getComponent)) {
        return callback(new Error('Provided source failed to create a runnable component'));
      }
      return this.registerComponent(packageId, name, implementation, function() {
        return callback(null);
      });
    };

    ComponentLoader.prototype.getSource = function(name, callback) {
      var component, componentName, nameParts;
      if (!this.ready) {
        this.listComponents((function(_this) {
          return function() {
            return _this.getSource(name, callback);
          };
        })(this));
        return;
      }
      component = this.components[name];
      if (!component) {
        for (componentName in this.components) {
          if (componentName.split('/')[1] === name) {
            component = this.components[componentName];
            name = componentName;
            break;
          }
        }
        if (!component) {
          return callback(new Error("Component " + name + " not installed"));
        }
      }
      if (typeof component !== 'string') {
        return callback(new Error("Can't provide source for " + name + ". Not a file"));
      }
      nameParts = name.split('/');
      if (nameParts.length === 1) {
        nameParts[1] = nameParts[0];
        nameParts[0] = '';
      }
      if (this.isGraph(component)) {
        nofloGraph.loadFile(component, function(graph) {
          if (!graph) {
            return callback(new Error('Unable to load graph'));
          }
          return callback(null, {
            name: nameParts[1],
            library: nameParts[0],
            code: JSON.stringify(graph.toJSON()),
            language: 'json'
          });
        });
        return;
      }
      return fs.readFile(component, 'utf-8', function(err, code) {
        if (err) {
          return callback(err);
        }
        return callback(null, {
          name: nameParts[1],
          library: nameParts[0],
          language: utils.guessLanguageFromFilename(component),
          code: code
        });
      });
    };

    ComponentLoader.prototype.readPackage = function(packageId, callback) {
      return this.getPackagePath(packageId, (function(_this) {
        return function(err, packageFile) {
          if (err) {
            return callback(err);
          }
          if (!packageFile) {
            return callback(new Error('no package found'));
          }
          return _this.readPackageFile(packageFile, callback);
        };
      })(this));
    };

    ComponentLoader.prototype.readPackageFile = function(packageFile, callback) {
      return fs.readFile(packageFile, 'utf-8', function(err, packageData) {
        var data;
        if (err) {
          return callback(err);
        }
        data = JSON.parse(packageData);
        data.realPath = path.dirname(packageFile);
        return callback(null, data);
      });
    };

    ComponentLoader.prototype.writePackage = function(packageId, data, callback) {
      return this.getPackagePath(packageId, function(err, packageFile) {
        var packageData;
        if (err) {
          return callback(err);
        }
        if (!packageFile) {
          return callback(new Error('no package found'));
        }
        if (data.realPath) {
          delete data.realPath;
        }
        packageData = JSON.stringify(data, null, 2);
        return fs.writeFile(packageFile, packageData, callback);
      });
    };

    ComponentLoader.prototype.registerComponentToDisk = function(packageId, name, cPath, callback) {
      if (callback == null) {
        callback = function() {};
      }
      return this.readPackage(packageId, (function(_this) {
        return function(err, packageData) {
          if (err) {
            return callback(err);
          }
          if (!packageData.noflo) {
            packageData.noflo = {};
          }
          if (!packageData.noflo.components) {
            packageData.noflo.components = {};
          }
          packageData.noflo.components[name] = cPath;
          _this.clear();
          return _this.writePackage(packageId, packageData, callback);
        };
      })(this));
    };

    ComponentLoader.prototype.registerGraphToDisk = function(packageId, name, cPath, callback) {
      if (callback == null) {
        callback = function() {};
      }
      return this.readPackage(packageId, (function(_this) {
        return function(err, packageData) {
          if (err) {
            return callback(err);
          }
          if (!packageData.noflo) {
            packageData.noflo = {};
          }
          if (!packageData.noflo.graphs) {
            packageData.noflo.graphs = {};
          }
          packageData.noflo.graphs[name] = cPath;
          _this.clear();
          return _this.writePackage(packageId, packageData, callback);
        };
      })(this));
    };

    return ComponentLoader;

  })(loader.ComponentLoader);

  exports.ComponentLoader = ComponentLoader;

}).call(this);
